### PR TITLE
executor: cache column info for prepare statement

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve? 

- cache column info arrays for prepare statement to avoid repeatedly doing column info generation
- ~~cache column info bytes for prepare statement to improve performance~~
 

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported variable/fields change

Side effects

 - Increased code complexity

Related changes

Release note

### CODE REVIEW

File: server/conn.go [hunk 1]
```diff
@@ -1312,16 +1312,51 @@ func (cc *clientConn) writeResultset(ctx context.Context, rs ResultSet, binary b
 	return cc.flush()
 }
 
-func (cc *clientConn) writeColumnInfo(columns []*ColumnInfo, serverStatus uint16) error {
+func (cc *clientConn) writeColumnInfo(rs ResultSet, serverStatus uint16) error {
+	var err error
 	data := cc.alloc.AllocWithLen(4, 1024)
```
Comment (coocood) [inline server/conn.go]: We can move this line after 1336

File: server/conn.go [hunk 2]
```diff
@@ -1312,16 +1312,51 @@ func (cc *clientConn) writeResultset(ctx context.Context, rs ResultSet, binary b
...
-		if err := cc.writePacket(data); err != nil {
+		storer(data[4:])
+		if err = cc.writePacket(data); err != nil {
+			return err
+		}
+	}
+	return cc.writeEOF(serverStatus)
+}
+
+func (cc *clientConn) writeColumnInfoByBytes(data []byte, columnsInfoBytes [][]byte, serverStatus uint16) error {
```
Comment (coocood) [inline server/conn.go]: We don't need the `data` if we reserve 4 bytes at the beginning of the columnInfoBytes.

File: server/conn.go [hunk 3]
```diff
@@ -1312,16 +1312,51 @@ func (cc *clientConn) writeResultset(ctx context.Context, rs ResultSet, binary b
...
-	if err := cc.writePacket(data); err != nil {
+	storer(data[4:])
+	if err = cc.writePacket(data); err != nil {
 		return err
 	}
 	for _, v := range columns {
 		data = data[0:4]
 		data = v.Dump(data)
-		if err := cc.writePacket(data); err != nil {
+		storer(data[4:])
```
Comment (coocood) [inline server/conn.go]: We can store `data` with leading 4 bytes.

File: server/conn.go [hunk 4]
```diff
@@ -1312,16 +1312,51 @@ func (cc *clientConn) writeResultset(ctx context.Context, rs ResultSet, binary b
...
+		if err = cc.writePacket(data); err != nil {
```
Comment (coocood) [inline server/conn.go]: We can reset the `ps.ColumnsInfoBytes` here instead of in the defer function.

File: server/conn.go [hunk 5]
```diff
@@ -1312,16 +1312,51 @@ func (cc *clientConn) writeResultset(ctx context.Context, rs ResultSet, binary b
...
+		}
+	}()
+	// try to use cached column info bytes writing packet directly
+	if ps != nil && ps.ColumnsInfoBytes != nil {
+		return cc.writeColumnInfoByBytes(data, ps.ColumnsInfoBytes, serverStatus)
+	}
+	columns := rs.Columns()
 	data = dumpLengthEncodedInt(data, uint64(len(columns)))
-	if err := cc.writePacket(data); err != nil {
+	storer(data[4:])
```
Comment (coocood) [inline server/conn.go]: We don't need to store the length packet, the length is the length of `ColumnsInfoBytes` so we can inline `storer` function into line 1345.

File: server/conn.go [hunk 6]
```diff
@@ -1312,16 +1312,45 @@ func (cc *clientConn) writeResultset(ctx context.Context, rs ResultSet, binary b
...
+func (cc *clientConn) writeColumnInfo(rs ResultSet, serverStatus uint16) error {
+	var err error
+	ps := rs.PrepareStmt()
+	columns := rs.Columns()
 	data := cc.alloc.AllocWithLen(4, 1024)
+	// send length packet
 	data = dumpLengthEncodedInt(data, uint64(len(columns)))
-	if err := cc.writePacket(data); err != nil {
+	if err = cc.writePacket(data); err != nil {
+		if ps != nil {
```
Comment (coocood) [inline server/conn.go]: We don't need to do it here.

**Comment by cfzjywxk** [comment]
`rs.Columns()` is an expensive call, try to skip this usage

File: server/driver_tidb.go [hunk 7]
```diff
@@ -351,10 +353,15 @@ func (tc *TiDBContext) GetSessionVars() *variable.SessionVars {
...
-	rows      []chunk.Row
-	closed    int32
+	recordSet    sqlexec.RecordSet
+	columns      []*ColumnInfo
+	rows         []chunk.Row
+	closed       int32
+	preparedStmt *core.CachedPrepareStmt
+}
+
+func (trs *tidbResultSet) SetPs(ps *core.CachedPrepareStmt) {
```
Comment (coocood) [inline server/driver_tidb.go]: This function is not used.

**Comment by coocood** [comment]
@cfzjywxk 
We don't need to store ColumnInfo for text protocol, text prepared statement it's rarely used.

**Comment by cfzjywxk** [comment]
> @cfzjywxk
> We don't need to store ColumnInfo for text protocol, text prepared statement it's rarely used.

`prepareStatement` is only  saved in `TiDBStatement.Execute` which will only be used by binary protocol

File: server/conn.go [hunk 8]
```diff
@@ -1344,23 +1378,25 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
...
+		}
 		rowCount := req.NumRows()
 		if rowCount == 0 {
 			break
 		}
 		for i := 0; i < rowCount; i++ {
 			data = data[0:4]
 			if binary {
-				data, err = dumpBinaryRow(data, rs.Columns(), req.GetRow(i))
+				data, err = dumpBinaryRow(data, columnInfo, req.GetRow(i))
```
Comment (coocood) [inline server/conn.go]: We can call `rs.Columns()` directly since it's already cached in rs.

File: planner/core/cache.go [hunk 9]
```diff
@@ -138,6 +138,8 @@ func NewPSTMTPlanCacheValue(plan Plan, names []*types.FieldName) *PSTMTPlanCache
 
 // CachedPrepareStmt store prepared ast from PrepareExec and other related fields
 type CachedPrepareStmt struct {
-	PreparedAst *ast.Prepared
-	VisitInfos  []visitInfo
+	PreparedAst      *ast.Prepared
+	VisitInfos       []visitInfo
+	ColumnsInfoBytes [][]byte
```
Comment (coocood) [inline planner/core/cache.go]: We can store the bytes directly in `ColumnInfo`.

File: server/driver_tidb.go [hunk 10]
```diff
@@ -391,12 +394,30 @@ func (trs *tidbResultSet) OnFetchReturned() {
...
+			}
+		}
+	}
 	if trs.columns == nil {
 		fields := trs.recordSet.Fields()
 		for _, v := range fields {
 			trs.columns = append(trs.columns, convertColumnInfo(v))
 		}
 	}
+	if trs.preparedStmt != nil {
```
Comment (coocood) [inline server/driver_tidb.go]: We can move this into `if trs.columns == nil`, then we can remove line 405-407

File: server/conn.go [hunk 11]
```diff
@@ -1313,15 +1313,16 @@ func (cc *clientConn) writeResultset(ctx context.Context, rs ResultSet, binary b
 }
 
 func (cc *clientConn) writeColumnInfo(columns []*ColumnInfo, serverStatus uint16) error {
+	var err error
 	data := cc.alloc.AllocWithLen(4, 1024)
 	data = dumpLengthEncodedInt(data, uint64(len(columns)))
-	if err := cc.writePacket(data); err != nil {
+	if err = cc.writePacket(data); err != nil {
```
Comment (coocood) [inline server/conn.go]: We don't need to modify this line now.

File: server/conn.go [hunk 12]
```diff
@@ -1344,8 +1345,7 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
 		if !gotColumnInfo {
 			// We need to call Next before we get columns.
 			// Otherwise, we will get incorrect columns info.
-			columns := rs.Columns()
-			err = cc.writeColumnInfo(columns, serverStatus)
+			err = cc.writeColumnInfo(rs.Columns(), serverStatus)
```
Comment (coocood) [inline server/conn.go]: This modify can be removed.

**Comment by coocood** [comment]
LGTM

**Comment by cfzjywxk** [comment]
@jackysp  @lysu  PTAL

File: server/driver_tidb.go [hunk 13]
```diff
@@ -391,11 +394,26 @@ func (trs *tidbResultSet) OnFetchReturned() {
...
+		return trs.columns
+	}
+	// for prepare statement, try to get cached columnInfo array
+	if trs.preparedStmt != nil {
+		ps := trs.preparedStmt
+		if colInfos, ok := ps.ColumnInfos.([]*ColumnInfo); ok {
+			trs.columns = colInfos
+		}
+	}
 	if trs.columns == nil {
```
Comment (lysu) [inline server/driver_tidb.go]: we can remove this `if`~?

Comment (coocood) [reply server/driver_tidb.go]: No, when we cached info, but `Columns` for a statement is first called, `trs.columns` will not be nil at this line.

**Pull request merged**